mm/mm_heap/mm_free.c

```diff
@@ -83,39 +83,16 @@ void mm_free(FAR struct mm_heap_s *heap, FAR void *mem)
       return;
     }
 
-#if defined(CONFIG_BUILD_FLAT) || defined(__KERNEL__)
-  /* Check current environment */
-
-  if (up_interrupt_context())
-    {
```

Comment:
@xiaoxiang781216  Where  is the equivalent test for the up_interrupt_context() to do a delayed action?

Reply:
The memory will free automatically in all allocatinon function:
https://github.com/apache/incubator-nuttx/pull/4096/files#diff-35207c436ed9a9af44c9ac830e37df62e859142009ec7b09a24f81fb37caf4a7L107

Reply:
The way I read this was: Under an interrupt context use mm_add_delaylist(heap, mem) to delay free the memory. 

Now what happens?

Reply:
The memory will be freed in the next allocation(e.g. malloc/realloc/zalloc/memalign...).

Reply:
I think what David is asking is where are we know adding the memory to the delayed list when we are in an interrupt context. Is `mm_takesemaphore() == false` enough?

Reply:
free should never be called in the interrupt context, so I add DEBUGASSERT(up_interrupt_context() == false) in mm_takesemaphore and remove the check here.

Reply:
> free should never be called in the interrupt context, so I add DEBUGASSERT(up_interrupt_context() == false) in mm_takesemaphore and remove the check here.

@xiaoxiang781216 

This worries me. I have a recollection of the delayed free coming in. (it broke large start ups scripts) 

There had to be a reason it was there? Have you asked the original author?

Reply:
> > free should never be called in the interrupt context, so I add DEBUGASSERT(up_interrupt_context() == false) in mm_takesemaphore and remove the check here.
> 
> @xiaoxiang781216
> 
> This worries me. I have a recollection of the delayed free coming in. (it broke large start ups scripts)
> 
> There had to be a reason it was there?

The comment already mention the reason why we can't free the memory immediately(most case is we can't free the running stack by the exiting thread self).

>  Have you asked the original author?

Ask for what?

But, anyway this patch doesn't modify the behaviour of delay list, what's this patch really change is remove the support of recurive lock since it never happen and then don't need.

Reply:
If "free should never be called in the interrupt context, so I add DEBUGASSERT(up_interrupt_context() == false)" is a true statement then explain why the [original code]( https://github.com/apache/incubator-nuttx/commit/1d2396353e5775db50794c59201731baf2dcc153 ) tested for it an used the delay queue. 

The PR breaks that behavior. I am asking you to verify if that was a bad change or this one is.

If we supporting free from up_interrupt_context (like post) this will be a breaking change to anyone that used it.

Reply:
Ok, mm_free is restored to support the interrupt context. Please take a look.

Reply:
Did you push?